ast/compile_test.go

```diff
@@ -664,6 +664,20 @@ p { true }`,
 		"mod6.rego": `package badrules.existserr
 
 p { true }`,
+		"mod7.rego": `package badrules.redeclaration
+
+p1 := 1
+p1 := 2
+
+p2 = 1
+p2 := 2`,
+		"mod8.rego": `package badrules.redeclaration_default
+
+default p3 := 1
+p3 := 2
```

Comment:
Functionally there doesn't seeem to be much difference anymore between `default p3 := 1` and `p3 := 1` right? Playing around with some examples it isn't clear how someone would be able to use the `:=` with `default` rules and still get the like fallback default behavior. Any rule that tries to define it to a different value would create a conflict/raise an error.

For some context on what I'm trying (on the very reasonable chance i'm missing something 😃)
```
{12:23} ✘ ~/p/g/s/g/o/opa:pr/1586 ◈ ❯ cat t.rego
package a

p3 := 1

p3 = x {
	x = input.x
}

{12:26} ~/p/g/s/g/o/opa:pr/1586 ◈ ❯ cat input.json
{
	"x": 7
}
{12:26} ~/p/g/s/g/o/opa:pr/1586 ◈ ❯ ./opa_darwin_amd64 eval -i ./input.json -d ./t.rego  "data"
{
  "error": [
    {
      "code": "rego_type_error",
      "message": "rule named p3 redeclared at ./t.rego:5",
      "location": {
        "file": "./t.rego",
        "row": 3,
        "col": 7
      }
    }
  ]
}
{12:26} ✘ ~/p/g/s/g/o/opa:pr/1586 ◈ ❯ vi t.rego
{12:27} ~/p/g/s/g/o/opa:pr/1586 ◈ ❯ cat t.rego
package a

default p3 := 1

p3 = x {
	x = input.x
}

{12:27} ~/p/g/s/g/o/opa:pr/1586 ◈ ❯ ./opa_darwin_amd64 eval -i ./input.json -d ./t.rego  "data"
{
  "error": [
    {
      "code": "rego_type_error",
      "message": "rule named p3 redeclared at ./t.rego:5",
      "location": {
        "file": "./t.rego",
        "row": 3,
        "col": 1
      }
    }
  ]
}
{12:27} ✘ ~/p/g/s/g/o/opa:pr/1586 ◈ ❯ vi t.rego
{12:27} ~/p/g/s/g/o/opa:pr/1586 ◈ ❯ cat t.rego
package a

default p3 = 1

p3 = x {
	x = input.x
}

{12:27} ~/p/g/s/g/o/opa:pr/1586 ◈ ❯ ./opa_darwin_amd64 eval -i ./input.json -d ./t.rego  "data"
{
  "result": [
    {
      "expressions": [
        {
          "value": {
            "a": {
              "p3": 7
            }
          },
          "text": "data",
          "location": {
            "row": 1,
            "col": 1
          }
        }
      ]
    }
  ]
}
```

I guess what this leads me to wondering is whether or not we should actually support using the assignment operator with default rules.. would it actually be the behavior someone expects? Maybe raising an error saying you can't use assignment for default rules might be better from a usability perspective.

Reply:
A similar situation would be:

```
p := x {
  # some conditions
}
```

If the conditions fail, p is undefined. If you can't declare a default value then you can't recover from which is not great. We could permit a default rule for each assignment. If we permit a default rule then we should probably also permit one or more else statements.

Reply:
I'm reading the semantics of := as basically: "here's an assignment of a variable and you NEVER need to look anywhere else to know what value gets assigned this variable."  Today we have that inside of rules; having it at the module level is pretty useful from a readability perspective.

Supporting `:=` within default seems at odds with that because `default` for me means "the value for this variable is really defined elsewhere but in case that doesn't work here's what it will end up as".  Prohibiting `:=` within `default` doesn't lose us anything; you get the same thing with raw `:=`.